replica/table.cc

```diff
@@ -594,26 +561,99 @@ void table::update_stats_for_new_sstable(const sstables::shared_sstable& sst) no
     _stats.live_sstable_count++;
 }
 
+void table::update_sstables_stats(ssize_t count, int64_t bytes_on_disk) noexcept {
+    _stats.live_disk_space_used += bytes_on_disk;
+    _stats.total_disk_space_used += bytes_on_disk;
+    _stats.live_sstable_count += count;
+}
+
+using is_main = bool_class<struct is_main_tag>;
+
+class table_sstables_adder : public row_cache::external_updater_impl {
+    table& _t;
+    compaction_group& _cg;
+    std::vector<sstables::shared_sstable> _sstables;
+    is_main _is_main;
+    lw_shared_ptr<memtable> _mut;
+
+    lw_shared_ptr<sstables::sstable_set> _new_sstable_set;
+    mutation_source_opt _ms_opt;
+public:
+    table_sstables_adder(table& t, compaction_group& cg, std::vector<sstables::shared_sstable> sstables, is_main main, lw_shared_ptr<memtable> m)
+        : _t(t)
+        , _cg(cg)
+        , _sstables(std::move(sstables))
+        , _is_main(main)
+        , _mut(std::move(m))
+    {
+        for (const auto& sst : _sstables) {
+            if (belongs_to_other_shard(sst->get_shards_for_this_sstable())) {
+                on_internal_error(tlogger, format("Attempted to load the shared SSTable {} at table", sst->get_filename()));
+            }
+        }
+        if (_mut) {
+            if (_sstables.size() == 1) {
+                _ms_opt = _sstables.front()->as_mutation_source();
+            } else {
+                std::vector<mutation_source> sources;
+                sources.reserve(_sstables.size());
+                for (auto& sst : _sstables) {
+                    sources.push_back(sst->as_mutation_source());
+                }
+                _ms_opt = make_combined_mutation_source(std::move(sources));
+            }
+        }
+    }
+
+    // Guarantees strong exception safety
+    virtual future<> prepare() override {
+        auto cur = _is_main ? _cg.main_sstables() : _cg.maintenance_sstables();
+        _new_sstable_set = make_lw_shared<sstables::sstable_set>(*cur);
+        auto inserted = _new_sstable_set->insert(_sstables);
```

Comment:
Reuse table::sstable_list_builder? It yields if needed. I know it's the same approach as before. I am thinking about LCS, where set being closed may contain thousands of elements to be copied in a single shot using sstable_set::clone().

Reply:
I think it's an overkill as it's doing too much and even the new_sstable_list (which is actually an sstable_set) needs to be passed by value, so here we'll need to create a sstable_set containing _sstables with no real reason.

A later patch also introduces the use of sstable_set::clone()
so it we're worried about preemption, we can add clone_gently that would yield.